File: npm/tests/doctor.test.js
'use strict'

// Mirrors internal/integrations/doctor_test.go and internal/commands/doctor_test.go.
// classifyDoctor is a pure function; runDoctor exercises the real
// IntegrationManager/buildPlans path so inspectResolved-equivalent logic is
// proven reused, not reimplemented.

const test = require('node:test')
const assert = require('node:assert/strict')
const fs = require('node:fs')
const os = require('node:os')
const path = require('node:path')

const { buildPlans, IntegrationManager } = require('../src/integrations')
const { classifyDoctor, runDoctor, UNREGISTERED_WRITE, HAND_MODIFIED, UNKNOWN_CONTENT } = require('../src/integrations/doctor')

function roots() {
  const base = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-doctor-'))
  const projectRoot = path.join(base, 'project')
  const homeRoot = path.join(base, 'home')
  fs.mkdirSync(projectRoot)
  fs.mkdirSync(homeRoot)
  return { projectRoot, homeRoot }
}

test('classifyDoctor: template match, no manifest entry -> unregistered write', () => {
  const status = { claim: { target: 'claude', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' }, destination: '/proj/x.md', state: 'current', managed: false, registered: false }
  const findings = classifyDoctor([status])
  assert.equal(findings.length, 1)
  assert.equal(findings[0].finding, UNREGISTERED_WRITE)
  assert.ok(findings[0].remedy.length > 0)
})

test('classifyDoctor: manifest-owned, hash differs -> hand modified', () => {
  const status = { claim: { target: 'claude', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' }, destination: '/proj/x.md', state: 'modified', managed: true, registered: true }
  const findings = classifyDoctor([status])
  assert.equal(findings.length, 1)
  assert.equal(findings[0].finding, HAND_MODIFIED)
})

// Was "not our problem" / length 0 before ML-2C — this is exactly the state
// ML-3A's audit found silently falling outside classifyDoctor's cases,
// which is what makes `agents install`'s preflight refuse with "unmanaged
// artifact". See UNKNOWN_CONTENT's doc comment.
test('classifyDoctor: content matches neither template nor manifest entry -> unknown content', () => {
  const status = { claim: { target: 'claude', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' }, destination: '/proj/x.md', state: 'modified', managed: false, registered: false }
  const findings = classifyDoctor([status])
  assert.equal(findings.length, 1)
  assert.equal(findings[0].finding, UNKNOWN_CONTENT)
  assert.ok(findings[0].remedy.includes('unmanaged artifact'))
})

test('classifyDoctor: template match, already registered and owned -> nothing to report', () => {
  const status = { claim: { target: 'claude', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' }, destination: '/proj/x.md', state: 'current', managed: true, registered: true }
  assert.equal(classifyDoctor([status]).length, 0)
})

test('classifyDoctor: registered under a DIFFERENT claim, content current -> must not be reported as unregistered', () => {
  const status = { claim: { target: 'claude', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' }, destination: '/proj/x.md', state: 'current', managed: false, registered: true }
  assert.equal(classifyDoctor([status]).length, 0)
})

// The unknown-content analogue of the case above: a destination registered
// under a DIFFERENT claim whose content also mismatches must stay silent
// too — it is that other claim's concern, not this one's, regardless of
// state. This is the discriminant Cenário 72 (check-gates-falsify.sh)
// falsifies for UNKNOWN_CONTENT, mirroring Cenário 71 for UNREGISTERED_WRITE.
test('classifyDoctor: registered under a DIFFERENT claim, content modified -> must not be reported as unknown content', () => {
  const status = { claim: { target: 'claude', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' }, destination: '/proj/x.md', state: 'modified', managed: false, registered: true }
  assert.equal(classifyDoctor([status]).length, 0)
})

test('classifyDoctor: sort key is total across a destination shared by two claims', () => {
  const base = { destination: '/proj/shared.md', state: 'current', managed: false, registered: false }
  const a = { ...base, claim: { target: 'zzz', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' } }
  const b = { ...base, claim: { target: 'aaa', surface: 'cli', scope: 'project', kind: 'agents', item: 'backend' } }
  const findings = classifyDoctor([a, b])
  assert.equal(findings.length, 2)
  assert.equal(findings[0].claim.target, 'aaa')
  assert.equal(findings[1].claim.target, 'zzz')
})

test('runDoctor: empty project reports zero findings', () => {
  const { projectRoot, homeRoot } = roots()
  const findings = runDoctor({ identity: { agents: {} }, projectRoot, homeRoot })
  assert.deepEqual(findings, [])
})

test('runDoctor: finds unregistered write after manifest entry removed, distinguishes it from a later hand edit', () => {
  const { projectRoot, homeRoot } = roots()
  const manager = new IntegrationManager({ projectRoot, homeRoot })
  const [plan] = buildPlans('agents', { targets: ['claude'], items: ['backend'], scope: 'project', identity: { agents: {} } })
  manager.install([plan])

  const manifestFile = manager.manifestPath('project')
  const manifest = JSON.parse(fs.readFileSync(manifestFile, 'utf8'))
  assert.equal(Object.keys(manifest.artifacts).length, 1)
  manifest.artifacts = {}
  fs.writeFileSync(manifestFile, JSON.stringify(manifest, null, 2))

  let findings = runDoctor({ identity: { agents: {} }, projectRoot, homeRoot })
  assert.equal(findings.length, 1)
  assert.equal(findings[0].finding, UNREGISTERED_WRITE)

  // Re-register normally, then hand-edit — must classify differently.
  manager.install([plan], { force: true })
  const destination = manager.resolve('project', plan.destination)
  fs.writeFileSync(destination, 'edited by hand')

  findings = runDoctor({ identity: { agents: {} }, projectRoot, homeRoot })
  assert.equal(findings.length, 1)
  assert.equal(findings[0].finding, HAND_MODIFIED)
})

test('runDoctor: finds unknown content for a destination never installed through trackfw', () => {
  const { projectRoot, homeRoot } = roots()
  const [plan] = buildPlans('agents', { targets: ['claude'], items: ['backend'], scope: 'project', identity: { agents: {} } })
  const destination = path.join(projectRoot, plan.destination)
  fs.mkdirSync(path.dirname(destination), { recursive: true })
  // Never installed (zero manifest entry), content matches neither the
  // catalog template nor any LegacyHashes entry.
  fs.writeFileSync(destination, 'nobody installed this through trackfw')

  const findings = runDoctor({ identity: { agents: {} }, projectRoot, homeRoot })
  assert.equal(findings.length, 1)
  assert.equal(findings[0].finding, UNKNOWN_CONTENT)
  assert.ok(findings[0].remedy.includes('unmanaged artifact'))
})

File: internal/integrations/doctor_test.go
package integrations

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/kgsaran/trackfw/internal/identity"
)

// TestClassifyDoctorDistinguishesTheThreeClasses exercises the seven cases
// the classification must get right without touching disk or the manifest —
// ClassifyDoctor only reads Inspection/PlannedArtifact fields. See the ML-2A
// roadmap entry for why classes 1 and 3 cannot be merged, ML-2C for why
// class 2 exists, and the doc comment on ClassifyDoctor for why Registered
// (not Managed) is the correct signal for classes 1 and 2.
func TestClassifyDoctorDistinguishesTheThreeClasses(t *testing.T) {
	plan := testPlan("project", "/proj/.claude/agents/trackfw-backend.md", "v1", "rendered")

	cases := []struct {
		name       string
		inspection Inspection
		wantKind   DoctorFindingKind
		wantFlag   bool
	}{
		{
			name: "template match, no manifest entry -> unregistered write",
			inspection: Inspection{
				Claim: plan.Claim, Destination: plan.Destination,
				State: StateCurrent, Managed: false, Registered: false,
			},
			wantKind: DoctorUnregisteredWrite, wantFlag: true,
		},
		{
			name: "manifest-owned, hash differs -> hand modified",
			inspection: Inspection{
				Claim: plan.Claim, Destination: plan.Destination,
				State: StateModified, Managed: true, Registered: true,
			},
			wantKind: DoctorHandModified, wantFlag: true,
		},
		{
			// Was "not our problem" / wantFlag: false before ML-2C — this is
			// exactly the state ML-3A's audit found silently falling outside
			// the switch, which is what makes `agents install`'s preflight
			// refuse with "unmanaged artifact". See DoctorUnknownContent's
			// doc comment.
			name: "content matches neither template nor manifest entry -> unknown content",
			inspection: Inspection{
				Claim: plan.Claim, Destination: plan.Destination,
				State: StateModified, Managed: false, Registered: false,
			},
			wantKind: DoctorUnknownContent, wantFlag: true,
		},
		{
			name: "template match, already registered and owned -> nothing to report",
			inspection: Inspection{
				Claim: plan.Claim, Destination: plan.Destination,
				State: StateCurrent, Managed: true, Registered: true,
			},
			wantFlag: false,
		},
		{
			name: "registered under a DIFFERENT claim, content current -> must not be reported as unregistered",
			inspection: Inspection{
				Claim: plan.Claim, Destination: plan.Destination,
				State: StateCurrent, Managed: false, Registered: true,
			},
			wantFlag: false,
		},
		{
			// The unknown-content analogue of the case above: a destination
			// registered under a DIFFERENT claim whose content also
			// mismatches must stay silent too — it is that other claim's
			// concern, not this one's, regardless of State. This is the
			// discriminant Cenário 72 (check-gates-falsify.sh) falsifies for
			// DoctorUnknownContent, mirroring Cenário 71 for
			// DoctorUnregisteredWrite.
			name: "registered under a DIFFERENT claim, content modified -> must not be reported as unknown content",
			inspection: Inspection{
				Claim: plan.Claim, Destination: plan.Destination,
				State: StateModified, Managed: false, Registered: true,
			},
			wantFlag: false,
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			findings := ClassifyDoctor([]PlannedArtifact{plan}, []Inspection{tc.inspection})
			if tc.wantFlag && len(findings) != 1 {
				t.Fatalf("findings = %d, want 1 (%+v)", len(findings), findings)
			}
			if !tc.wantFlag && len(findings) != 0 {
				t.Fatalf("findings = %d, want 0 (%+v)", len(findings), findings)
			}
			if tc.wantFlag {
				if findings[0].FindingKind != tc.wantKind {
					t.Fatalf("finding kind = %q, want %q", findings[0].FindingKind, tc.wantKind)
				}
				if findings[0].Remedy == "" {
					t.Fatal("remedy is empty")
				}
			}
		})
	}
}

// TestClassifyDoctorSortIsTotal proves the sort key is total across a
// destination shared by two claims (ML-2B's gate needs deterministic order
// across three independent CLI implementations).
func TestClassifyDoctorSortIsTotal(t *testing.T) {
	shared := "/proj/.agents/shared.md"
	planA := testPlan("project", shared, "v1", "content")
	planA.Claim.Target = "zzz"
	planB := testPlan("project", shared, "v1", "content")
	planB.Claim.Target = "aaa"

	inspection := func(p PlannedArtifact) Inspection {
		return Inspection{Claim: p.Claim, Destination: p.Destination, State: StateCurrent, Managed: false, Registered: false}
	}
	findings := ClassifyDoctor([]PlannedArtifact{planA, planB}, []Inspection{inspection(planA), inspection(planB)})
	if len(findings) != 2 {
		t.Fatalf("findings = %d, want 2", len(findings))
	}
	if findings[0].Claim.Target != "aaa" || findings[1].Claim.Target != "zzz" {
		t.Fatalf("unexpected order: %+v", findings)
	}
}

// TestRunDoctorFindsUnregisteredWriteAfterManifestEntryRemoved reproduces the
// state ML-1A's inversion was written to prevent going forward, but that can
// already exist on disk from before it landed: bytes on disk are byte-
// identical to the current catalog render, but the manifest has no entry for
// the destination. This is an end-to-end test through Manager (not
// hand-built Inspection values), proving inspectResolved/List really is
// reused, not reimplemented.
func TestRunDoctorFindsUnregisteredWriteAfterManifestEntryRemoved(t *testing.T) {
	manager, project, _ := testManager(t)
	plan := testPlan("project", ".claude/agents/trackfw-backend.md", "v1", "rendered content")

	if err := manager.Install([]PlannedArtifact{plan}, false); err != nil {
		t.Fatal(err)
	}

	// Simulate the pre-ADR-2026-08-18 crash window: bytes landed, manifest
	// entry for this destination did not.
	manifestFile := manifestPath(project)
	manifest, err := loadManifest(manifestFile)
	if err != nil {
		t.Fatal(err)
	}
	absoluteDestination := filepath.Join(project, plan.Destination)
	delete(manifest.Artifacts, absoluteDestination)
	if err := writeManifest(manifestFile, manifest); err != nil {
		t.Fatal(err)
	}

	findings := ClassifyDoctor([]PlannedArtifact{plan}, mustList(t, manager, plan))
	if len(findings) != 1 {
		t.Fatalf("findings = %d, want 1: %+v", len(findings), findings)
	}
	if findings[0].FindingKind != DoctorUnregisteredWrite {
		t.Fatalf("finding kind = %q, want %q", findings[0].FindingKind, DoctorUnregisteredWrite)
	}

	// Same destination but a genuine hand edit after a normal install must
	// classify differently, not collapse into the same bucket.
	if err := os.WriteFile(absoluteDestination, []byte("edited by hand"), 0o600); err != nil {
		t.Fatal(err)
	}
	// Re-register normally, so this second case starts from Registered+owned.
	if err := manager.Install([]PlannedArtifact{plan}, true); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(absoluteDestination, []byte("edited by hand again"), 0o600); err != nil {
		t.Fatal(err)
	}
	findings = ClassifyDoctor([]PlannedArtifact{plan}, mustList(t, manager, plan))
	if len(findings) != 1 || findings[0].FindingKind != DoctorHandModified {
		t.Fatalf("findings = %+v, want single hand-modified finding", findings)
	}
}

// TestRunDoctorFindsUnknownContentWhenNeitherRegisteredNorMatching reproduces
// the state ML-3A's audit found falling silently outside ClassifyDoctor's
// switch: a destination that was never installed (no manifest entry at all)
// but that carries content matching neither the catalog template nor any
// LegacyHashes entry. This is an end-to-end test through Manager (not
// hand-built Inspection values), proving inspectResolved really produces
// StateModified for this case (manager.go:638-645) and that ClassifyDoctor
// now has a case for it.
func TestRunDoctorFindsUnknownContentWhenNeitherRegisteredNorMatching(t *testing.T) {
	manager, project, _ := testManager(t)
	plan := testPlan("project", ".claude/agents/trackfw-backend.md", "v1", "rendered content")

	absoluteDestination := filepath.Join(project, plan.Destination)
	if err := os.MkdirAll(filepath.Dir(absoluteDestination), 0o755); err != nil {
		t.Fatal(err)
	}
	// Alien/orphaned bytes at the exact catalog destination, never installed
	// (zero manifest entry) — content matches neither `desired` nor any
	// LegacyHashes entry.
	if err := os.WriteFile(absoluteDestination, []byte("content nobody wrote through trackfw"), 0o600); err != nil {
		t.Fatal(err)
	}

	findings := ClassifyDoctor([]PlannedArtifact{plan}, mustList(t, manager, plan))
	if len(findings) != 1 {
		t.Fatalf("findings = %d, want 1: %+v", len(findings), findings)
	}
	if findings[0].FindingKind != DoctorUnknownContent {
		t.Fatalf("finding kind = %q, want %q", findings[0].FindingKind, DoctorUnknownContent)
	}
	if !strings.Contains(findings[0].Remedy, "unmanaged artifact") {
		t.Fatalf("remedy must name the preflight refusal literally, got: %q", findings[0].Remedy)
	}
}

// TestRunDoctorSweepsFullCatalogWithoutError proves the per-(target,surface)
// sweep in doctorPlansForScope tolerates surfaces that do not support a
// given scope (pathForScope returning ok=false) instead of failing the way a
// single BuildPlans(AllSurfaces: true, no Targets filter) call would — see
// RunDoctor's doc comment. An empty project must report zero findings.
func TestRunDoctorSweepsFullCatalogWithoutError(t *testing.T) {
	manager, _, _ := testManager(t)
	catalog, err := LoadCatalog()
	if err != nil {
		t.Fatal(err)
	}
	findings, err := RunDoctor(catalog, manager, identity.Config{}, nil)
	if err != nil {
		t.Fatalf("RunDoctor() error = %v", err)
	}
	if len(findings) != 0 {
		t.Fatalf("findings on empty project = %d, want 0: %+v", len(findings), findings)
	}
}

func mustList(t *testing.T, manager Manager, plans ...PlannedArtifact) []Inspection {
	t.Helper()
	inspections, err := manager.List(plans)
	if err != nil {
		t.Fatal(err)
	}
	return inspections
}
